README.md
# cifar wgan
This repo contains a trained example for generating cifar using wasserstien gan architecture. Though model is trained until 2500 epochs in order to improve its kernel density you can train the model for couple of more iterations. The project is just here to provide you a kickstart on how to generate RGB channel images using GANs architectures.

More information about it could be found on [link](https://arxiv.org/abs/1701.07875)

![alt Cifar WGAN](./cifarwasserstien.png "Cifar WGAN")

### requirements
- tensorflow >= 2.0
- pillow
- matplotlib

### Preparing virtual env
If you dont have venv installed in your system you can do it by using
```
sudo apt install python3-venv
```
Now create virtual enviornment for your application
```
python -m venv env
```

Activating the venv
- Ubuntu/Linux
```
source env/bin/activate
```

- Windows
```
.\env\Scripts\activate
```

### Downloading the dataset
```python
(X_train, _), (x_test, _) = tf.keras.datasets.cifar10.load_data()

train_dataset = tf.data.Dataset.from_tensor_slices((tf.cast(X_train, tf.float32))).map(lambda x: (x-127.5)/127.5).shuffle(50000).batch(256)
```

### How to train model
```
python3 main.py
```

### How to Generate Images?
Image generation can be carried out using two different methods, which are as follows:

#### Using shell
```bash
python -i generate_images.py

image_generator = CifarImagesGenerator()
image_generator.generate_images()
```

#### In a python file
```python
from generate_images import CifarImagesGenerator

image_generator = CifarImagesGenerator()
image_generator.generate_images()
```


changelog.md
### 2020-12-15
- added all the modules and training checkpoints
- added license, readme and generate image module


models.py
"""
Author: Sanidhya Mangal
Github: sanidhyamangal
"""
import tensorflow as tf  # for deep learning related steps


# create a sample model for the dcgan
class WasserstienGenerative(tf.keras.models.Model):
    """
    Class for implementaion of Generative model for Wasserstien GAN Arch
    """
    def __init__(self, *args, **kwargs):
        super(WasserstienGenerative, self).__init__(*args, **kwargs)

        self.model = tf.keras.Sequential([
            tf.keras.layers.Dense(8 * 8 * 256,
                                  use_bias=False,
                                  input_shape=(100, )),
            tf.keras.layers.BatchNormalization(),
            tf.keras.layers.LeakyReLU(),

            # reshape function
            tf.keras.layers.Reshape((8, 8, 256)),

            # first conv2d transpose
            tf.keras.layers.Conv2DTranspose(filters=128,
                                            kernel_size=(5, 5),
                                            strides=(1, 1),
                                            padding="same",
                                            use_bias=False),
            tf.keras.layers.BatchNormalization(),
            tf.keras.layers.LeakyReLU(),

            # second conv2d transpose
            tf.keras.layers.Conv2DTranspose(filters=64,
                                            kernel_size=(5, 5),
                                            strides=(2, 2),
                                            padding="same",
                                            use_bias=False),
            tf.keras.layers.BatchNormalization(),
            tf.keras.layers.LeakyReLU(),

            # third conv2d transpose
            tf.keras.layers.Conv2DTranspose(filters=3,
                                            kernel_size=(5, 5),
                                            strides=(2, 2),
                                            padding="same",
                                            use_bias=False,
                                            activation=tf.nn.tanh)
        ])

    def call(self, inputs):
        return self.model(inputs)


# create a sample model for the dcgan
class WasserstienDiscriminative(tf.keras.models.Model):
    """
    Discriminative Model for Wasserstien GAN Arch
    """
    def __init__(self, *args, **kwargs):
        super(WasserstienDiscriminative, self).__init__(*args, **kwargs)

        self.model = tf.keras.Sequential([
            tf.keras.layers.Conv2D(64,
                                   kernel_size=(5, 5),
                                   strides=(2, 2),
                                   padding="same"),
            tf.keras.layers.LeakyReLU(),
            tf.keras.layers.Dropout(0.3),

            # first conv2d transpose
            tf.keras.layers.Conv2D(filters=128,
                                   kernel_size=(5, 5),
                                   strides=(1, 1),
                                   padding="same"),
            tf.keras.layers.LeakyReLU(),
            tf.keras.layers.Dropout(0.3),

            # third conv2d transpose
            tf.keras.layers.Flatten(),
            tf.keras.layers.Dense(1)
        ])

    def call(self, inputs):
        return self.model(inputs)


generate_images.py
from models import WasserstienGenerative  # generative models
import tensorflow as tf  # for deep learning based ops
from utils import generate_and_save_images  # function to generate and save images

import os  # for os related ops


class CifarImagesGenerator:
    """
    Class for generating cifar images
    """
    def __init__(self,
                 path_to_checkpoint: str = './training_checkpoints',
                 *args,
                 **kwargs):
        self.generator = WasserstienGenerative()
        # creation of checkpoint dirs

        checkpoint_prefix = os.path.join(path_to_checkpoint, "ckpt")
        checkpoint = tf.train.Checkpoint(generator=self.generator)

        checkpoint.restore(tf.train.latest_checkpoint(checkpoint_dir))

    def generate_images(self, image_name: str = "generated"):
        seed = tf.random.normal([16, 100])
        generate_and_save_images(self.generator, image_name, seed)


main.py
"""
Author: Sanidhya Mangal
Github: sanidhyamangal
"""
import os  # for os related ops
import time  # for time related ops

import matplotlib.pyplot as plt  # for plotting
import PIL  # for image related ops
import tensorflow as tf  # for deep learning related steps
from IPython import display

from models import (
    WasserstienDiscriminative,  # load models
    WasserstienGenerative)
from utils import generate_and_save_images  # function to generate and save images

# Get traning dataset
(X_train, _), (x_test, _) = tf.keras.datasets.cifar10.load_data()

train_dataset = tf.data.Dataset.from_tensor_slices((tf.cast(
    X_train,
    tf.float32))).map(lambda x: (x - 127.5) / 127.5).shuffle(50000).batch(256)

# intansiate generator
generator = WasserstienGenerative()

print(generator.model.summary())
noise = tf.random.normal([1, 100])
generated_image = generator(noise, training=False)

plt.imshow(generated_image[0, :, :, 0])

discriminator = WasserstienDiscriminative()
decision = discriminator(generated_image)
print(decision)


# create a loss for generative
def generator_loss(fake_output):
    return -tf.reduce_mean(fake_output)


# create a loss function for discriminator
def discriminator_loss(real_output, fake_output):
    return tf.reduce_mean(fake_output) - tf.reduce_mean(real_output)


# optimizers for the generative models
generator_optimizer = tf.keras.optimizers.Adam(1e-4)
discriminator_optimizer = tf.keras.optimizers.Adam(1e-4)

# creation of checkpoint dirs
checkpoint_dir = './training_checkpoints'
checkpoint_prefix = os.path.join(checkpoint_dir, "ckpt")
checkpoint = tf.train.Checkpoint(
    generator_optimizer=generator_optimizer,
    discriminator_optimizer=discriminator_optimizer,
    generator=generator,
    discriminator=discriminator)

# all the epochs for the training
EPOCHS = 500
noise_dim = 100
num_examples_to_generate = 16
BATCH_SIZE = 256

# We will reuse this seed overtime (so it's easier)
# to visualize progress in the animated GIF)
seed = tf.random.normal([num_examples_to_generate, noise_dim])


@tf.function
def train_step(images):
    """
    TF Graph function to be compiled into a graph
    """
    noise = tf.random.normal([BATCH_SIZE, noise_dim])

    with tf.GradientTape() as gen_tape, tf.GradientTape() as disc_tape:
        generated_images = generator(noise, training=True)

        real_output = discriminator(images, training=True)
        fake_output = discriminator(generated_images, training=True)

        gen_loss = generator_loss(fake_output)
        disc_loss = discriminator_loss(real_output, fake_output)

    gradients_of_generator = gen_tape.gradient(gen_loss,
                                               generator.trainable_variables)
    gradients_of_discriminator = disc_tape.gradient(
        disc_loss, discriminator.trainable_variables)

    generator_optimizer.apply_gradients(
        zip(gradients_of_generator, generator.trainable_variables))
    discriminator_optimizer.apply_gradients(
        zip(gradients_of_discriminator, discriminator.trainable_variables))


def train(dataset, epochs):
    """
    Function to perform training ops on given set of epochs
    """
    for epoch in range(epochs):
        start = time.time()

        for image_batch in dataset:
            train_step(image_batch)

        # Produce images for the GIF as we go
        display.clear_output(wait=True)
        generate_and_save_images(generator, epoch + 1, seed)

        # Save the model every 15 epochs
        if (epoch + 1) % 100 == 0:
            checkpoint.save(file_prefix=checkpoint_prefix)

        print('Time for epoch {} is {} sec'.format(epoch + 1,
                                                   time.time() - start))

    # Generate after the final epoch
    display.clear_output(wait=True)
    generate_and_save_images(generator, epochs, seed)


train(train_dataset, EPOCHS)
